packages/yarnpkg-pnp/sources/loader/makeManager.ts

```diff
@@ -96,26 +96,39 @@ export function makeManager(pnpapi: PnpApi, opts: MakeManagerOptions) {
   }
 
   function findApiPathFor(modulePath: NativePath) {
-    const controlledBy: Array<PortablePath> = [];
+    const candidates: Array<{packageLocation: NativePath, apiPath: PortablePath}> = [];
+
     for (const [apiPath, apiEntry] of apiMetadata) {
       const locator = apiEntry.instance.findPackageLocator(modulePath);
 
-      if (locator) {
-        if (apiMetadata.size === 1) {
-          return apiPath;
-        } else {
-          controlledBy.push(apiPath);
-        }
-      }
+      if (!locator)
+        continue;
+
+      if (apiMetadata.size === 1)
+        return apiPath;
+
+      const packageInformation = apiEntry.instance.getPackageInformation(locator);
+      if (!packageInformation)
+        throw new Error(`Assertion failed: Couldn't get package information for '${modulePath}'`);
+
+      candidates.push({packageLocation: packageInformation.packageLocation, apiPath});
     }
 
-    if (controlledBy.length !== 0) {
-      if (controlledBy.length === 1)
-        return controlledBy[0];
+    if (candidates.length === 1)
+      return candidates[0].apiPath;
+
+    if (candidates.length > 1) {
+      const [candidateA, candidateB] = [...candidates].sort(
+        (a, b) => b.packageLocation.length - a.packageLocation.length
+      );
```

Comment:
Rather than doing a post-process `sort`, we could just keep track of the "best" `packageLocation` in the initial iteration, only updating the apiPath to return if the new value is longer than the current one.